src/naive_bayes/mod.rs

```diff
@@ -0,0 +1,239 @@
+    coef: Vec<Vec<Vec<T>>>,
+    feature_categories: Vec<Vec<T>>,
+}
+
+impl<T: RealNumber, M: Matrix<T>> NBDistribution<T, M> for CategoricalNB<T> {
+    fn prior(&self, k: T) -> T {
+        match self.class_labels.iter().position(|&t| t == k) {
+            Some(idx) => self.class_probabilities[idx],
+            None => T::zero(),
+        }
+    }
+
+    fn conditional_probability(&self, k: T, j: &M::RowVector) -> T {
+        match self.class_labels.iter().position(|&t| t == k) {
+            Some(idx) => {
+                let mut prob = T::one();
+                for feature in 0..j.len() {
+                    let value = j.get(feature);
+                    match self.feature_categories[feature]
+                        .iter()
+                        .position(|&t| t == value)
+                    {
+                        Some(_i) => prob *= self.coef[idx][feature][_i],
+                        None => return T::zero(),
+                    }
+                }
+                prob
+            }
+            None => T::zero(),
+        }
+    }
+
+    fn classes(&self) -> &Vec<T> {
+        &self.class_labels
+    }
+}
+
+impl<T: RealNumber> CategoricalNB<T> {
+    /// Fits the distribution to a NxM matrix where N is number of samples and M is number of features.
+    /// * `x` - training data.
+    /// * `y` - vector with target values (classes) of length N.
+    /// * `alpha` - Additive (Laplace/Lidstone) smoothing parameter (0 for no smoothing).
+    pub fn fit<M: Matrix<T>>(x: &M, y: &M::RowVector, alpha: T) -> Result<Self, Failed> {
+        if alpha < T::zero() {
+            return Err(Failed::fit(&format!(
+                "alpha should be >= 0, alpha=[{}]",
+                alpha
+            )));
+        }
+
+        let (n_samples, n_features) = x.shape();
+        let y_samples = y.len();
+        if y_samples != n_samples {
+            return Err(Failed::fit(&format!(
+                "Size of x should equal size of y; |x|=[{}], |y|=[{}]",
+                n_samples, y_samples
+            )));
+        }
+
+        if n_samples == 0 {
+            return Err(Failed::fit(&format!(
+                "Size of x and y should greater than 0; |x|=[{}]",
+                n_samples
+            )));
+        }
+
+        let mut y_sorted = y.to_vec();
+        y_sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
+        let mut class_labels = Vec::with_capacity(y.len());
+        class_labels.push(y_sorted[0]);
+        let mut classes_count = Vec::with_capacity(y.len());
+        let mut current_count = T::one();
+        for idx in 1..y_samples {
+            if y_sorted[idx] == y_sorted[idx - 1] {
+                current_count += T::one();
+            } else {
+                classes_count.push(current_count);
+                class_labels.push(y_sorted[idx]);
+                current_count = T::one()
+            }
+            classes_count.push(current_count);
+        }
+
+        let x = x.transpose();
```

Comment:
Why do you need to do transpose?

Reply:
not needed, I can do the same with get_col(). Removing